### PR TITLE
Implement AddAppointmentCommand

### LINKED ISSUE #77: Implement AddAppointment command

### PR DESCRIPTION
**Pull request by mounilsankar**
Closes #77 
1. Storage logic to be implemented in a future pull request in NewLogicManager
2. Integration test with NewModelManager also to be done

### CODE REVIEW

File: src/main/java/seedu/address/model/appointment/AppointmentEndTime.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Represents a Appointment's end time in the database.
+ * Guarantees: immutable; is valid as declared in {@link #isValidAppointmmentTime(String)}
+ */
+public class AppointmentEndTime {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Appointment end date and time should be of the format YYYY-MM-DD HHmm ";
+    public static final String DATETIME_INPUT_FORMAT = "yyyy-MM-dd HH:mm";
```
Comment (derekjxtan) [inline src/main/java/seedu/address/model/appointment/AppointmentEndTime.java:18]: doesnt match the format in MESSAGE_CONSTRAINTS above

File: src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        AddAppointmentCommand addAppointmentCommand = new AddAppointmentCommand(validAppointment);
+        ModelStub modelStub = new ModelStubWithAppointment(validAppointment);
+
+        assertThrows(CommandException.class,
+                AddAppointmentCommand.MESSAGE_DUPLICATE_APPOINTMENT, () -> addAppointmentCommand.execute(modelStub));
+    }
+
+    @Test
+    public void equals() {
+        Appointment appointment1 = new AppointmentBuilder().withPatientNric("T1234567N").build();
```
Comment (derekjxtan) [inline src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java:62]: should make use of the constants in TypicalPatients

File: src/test/java/seedu/address/logic/parser/AddAppointmentCommandParserTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+        // invalid phone
+        assertParseFailure(parser, INVALID_APPOINTMENT_START_TIME_DESC + validExpectedAppointmentString,
+                Messages.getErrorMessageForDuplicatePrefixes(PREFIX_APPOINTMENT_START_TIME));
+
+        // invalid address
+        assertParseFailure(parser, INVALID_APPOINTMENT_END_TIME_DESC + validExpectedAppointmentString,
+                Messages.getErrorMessageForDuplicatePrefixes(PREFIX_APPOINTMENT_END_TIME));
+
+        // valid value followed by invalid value
```
Comment (derekjxtan) [inline src/test/java/seedu/address/logic/parser/AddAppointmentCommandParserTest.java:101]: missing testcase here?

File: src/test/java/seedu/address/logic/parser/NewAddressBookParserTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import seedu.address.testutil.PatientBuilder;
+import seedu.address.testutil.PersonUtil;
+
+
+public class NewAddressBookParserTest {
+
+    private final NewAddressBookParser parser = new NewAddressBookParser();
+
+    @Test
+    public void parseCommand_addAppointment() throws Exception {
```
Comment (derekjxtan) [inline src/test/java/seedu/address/logic/parser/NewAddressBookParserTest.java:26]: test functions need to throw exceptions?

File: src/test/java/seedu/address/model/appointment/AppointmentEndTimeTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    @Test
+    public void constructor_invalidAppointmentTime_throwsIllegalArgumentException() {
+        String invalidName = "";
+        assertThrows(IllegalArgumentException.class, () -> new AppointmentEndTime(invalidName));
+    }
+
+    @Test
+    public void isValidAppointmentStartTime() {
+        // null name
+        assertThrows(NullPointerException.class, () -> AppointmentStartTime.isValidAppointmmentTime(null));
```
Comment (derekjxtan) [inline src/test/java/seedu/address/model/appointment/AppointmentEndTimeTest.java:25]: AppointmentEndTime

**Review by derekjxtan — COMMENTED** [review]
Any intention to test the NewLogicManager in this pr?

File: src/test/java/seedu/address/logic/parser/AddAppointmentCommandParserTest.java (hunk 3 above)
Comment (mounilsankar) [reply src/test/java/seedu/address/logic/parser/AddAppointmentCommandParserTest.java:101]: Its meant to describe the next set of test cases

File: src/test/java/seedu/address/logic/parser/NewAddressBookParserTest.java (hunk 4 above)
Comment (mounilsankar) [reply src/test/java/seedu/address/logic/parser/NewAddressBookParserTest.java:26]: the existing AddressBookParser test had it so i just followed

File: src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        if (!model.hasDoctorWithNric(toAdd.getDoctorNric())) {
+            throw new CommandException(MESSAGE_INVALID_DOCTOR);
+        }
+
+        if (!model.hasPatientWithNric(toAdd.getPatientNric())) {
+            throw new CommandException(MESSAGE_INVALID_PATIENT);
+        }
+
+        if (model.hasAppointment(toAdd)) {
```
[the remark this answers is not in the document]
Comment (mounilsankar) [reply src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java:57]: But the error mesage will be the same correct for appointment/doctor/patient if we use the checking by uniqueItemList unless try catch?

[the remark this answers is not in the document]
Comment (derekjxtan) [reply src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java:57]: should be fixed in UniqueItemList. Doesnt make sense to check for duplicates twice.

**Review by derekjxtan — APPROVED** [review]

**Pull request merged**